variable/type_conversion.cpp

```diff
@@ -14,22 +14,21 @@ namespace scipp::variable {
 struct MakeVariableWithType {
   template <class T> struct Maker {
     static Variable apply(const VariableConstView &parent) {
-      auto conversion = overloaded{
-          [](const units::Unit &x) { return x; },
-          [](const auto &x) {
-            if constexpr (is_ValueAndVariance_v<std::decay_t<decltype(x)>>)
-              return ValueAndVariance<T>{static_cast<T>(x.value),
-                                         static_cast<T>(x.variance)};
-            else
-              return static_cast<T>(x);
-          }};
-      if constexpr (core::canHaveVariances<T>())
-        return transform<double, float, int64_t, int32_t, bool>(parent,
-                                                                conversion);
-      else
-        return transform<double, float, int64_t, int32_t, bool>(
-            parent, overloaded{core::transform_flags::expect_no_variance_arg<0>,
-                               conversion});
+      using namespace core::transform_flags;
+      constexpr auto expect_input_variances =
+          enable_flag_if<!core::canHaveVariances<T>(),
+                         expect_no_variance_arg_t<0>>{};
```

Comment:
Yes

Reply:
I wish there was a `constexpr` ternary:
```cpp
constexpr auto expect_input_variances = canHaveVariances<T> constexpr ? null_flag : expect_no_variances_arg<0>;
```
Could do the same with a lambda but that does not look any nicer than using `enable_flag_if`.

But maybe `enable_flag_if` could be replaced by a `constexpr` function/lambda?

```cpp
constexpr auto conditional_flag = [](auto flag, const bool condition) {
  if constexpr(condition)
    return flag;
  else
   return null_flag;
};

// later
overloaded{flag_if(expect_no_variance_arg<0>, !core::canHaveVariances<T>()), ...}
```

Reply:
> I wish there was a constexpr ternary:
Yes, I'd like to have it and I've seen a number of posts requesting that feature too.

I suspect the constexpr lambda you propose above would work with the flags the way that they are. I think this `struct` approach is slightly easier to reason and work with.

Reply:
Not suggesting to change the struct approach, but instead of having to mess with both the flag classes and flag instances when using `enable_flag_if`, using a `constexpr` function instead of a struct would provide a more natural way. Users of transform would then not need to know that there is `some_flag` as well as `some_flag_t`.

Reply:
Agreed. Only working with the instances would make life easier.

Reply:
I think the best you can do is along the lines of this...
```
template<bool B>
constexpr auto conditional_flag = [](auto flag) {
   return std::conditional_t<B, decltype(flag), Flag>{};
};
```
With usage like:
```
conditional_flag<!core::canHaveVariances<T>()>(expect_no_variance_arg<0>);
```

Reply:
Are you saying the `constexpr` lambda I listed above does not work?

Reply:
Yes

Reply:
> I think the best you can do is along the lines of this...
> 
> ```c++
> template<bool B>
> constexpr auto conditional_flag = [](auto flag) {
>    return std::conditional_t<B, decltype(flag), Flag>{};
> };
> ```
> 
> With usage like:
> 
> ```c++
> conditional_flag<!core::canHaveVariances<T>()>(expect_no_variance_arg<0>);
> ```

This looks better to me 👍